### PR TITLE
WIP (SIMP-3062) Hook FakeCA to Puppet Client Certs

### PR DESCRIPTION
**Pull request by trevor-vaughan**
This adds an incron job to trigger the creation of a FakeCA certificate
for a newly minted client in the appropriate environment.

You *must* have the associated simp-core commit that makes the FakeCA
thread safe.

SIMP-3062 #close

### CODE REVIEW

File: files/usr/local/sbin/simp_fakeca_incron_hook [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  end
+elsif action == 'IN_DELETE'
+  keydist_path = '/var/simp/environments/simp/site_files/pki_files/files/keydist'
+
+  to_delete = File.join(keydist_path, fqdn)
+  backup_path = File.join(keydist_path, '.incron_backup')
+
+  if File.exist?(to_delete)
+    FileUtils.mkdir_p(backup_path)
+
```
Comment (lnemsick-simp) [inline files/usr/local/sbin/simp_fakeca_incron_hook]: Does it make sense to log this action?

File: files/usr/local/sbin/simp_fakeca_incron_hook [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  unless File.exist?(fakeca)
+    $stderr.puts "Error: No FakeCA found at '#{fakeca}'"
+    $stderr.puts "  Are you sure your environment is set up properly?"
+    exit 1
+  end
+
+  fakeca_cmd = %{#{fakeca} auto #{host_pki_id.join(',')}}
+
+  fakeca_cmd_output = %x{#{fakeca_cmd}}
+
```
Comment (lnemsick-simp) [inline files/usr/local/sbin/simp_fakeca_incron_hook]: Does it make sense to log success?

File: files/usr/local/sbin/simp_fakeca_incron_hook [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  unless File.exist?(fakeca)
+    $stderr.puts "Error: No FakeCA found at '#{fakeca}'"
+    $stderr.puts "  Are you sure your environment is set up properly?"
+    exit 1
+  end
+
+  fakeca_cmd = %{#{fakeca} auto #{host_pki_id.join(',')}}
+
+  fakeca_cmd_output = %x{#{fakeca_cmd}}
```
Comment (lnemsick-simp) [inline files/usr/local/sbin/simp_fakeca_incron_hook]: Won't capture stderr. So, if you want fakeca_cmd_output for errors and errors are on stderr, fakeca_cmd_output content will not help you.

**Comment by kendall-moore** [comment]
This code does not appear to work in a couple of circumstances which seems to me warrant refactoring. Here are those circumstances.

### Static Catalogs are Enabled

In PE, if static catalogs are enabled, file resources who have a remote source are cached and checksum'd so as to be able to check if a new catalog is needed or not in the future.  Because of this, the code fails at compilation time instead of application time causing the FakeCA signing to fail due to not being able to compile a bug.  This is arguably a bug in the implementation of static catalogs but either way blocks this code's ability to function.

### Compile Masters are Present

The simp_fakeca_incron_hook script appears to require that an FQDN is passed to it, which is the host that will have a catalog compiled for and information will be gathered for it.  The way that the incron script is created an FQDN is not passed as an argument, but rather the fully qualified path to the signed certificates directory inside of Puppet's $ssldir/ca directory.  This means that this can't work on compile masters without the ability for compile masters to broker this request to a CA and also then have the resulting signed certificates sync'd back to the compile masters.  Currently there is no such architecture built into and supported by SIMP.

**Comment by trevor-vaughan** [comment]
Closing this as not worth doing.

**Pull request closed without merging**